### PR TITLE
Traffic Ops Golang Incremental Rewrite App

### PR DESCRIPTION
**Pull request by rob05c**
This adds an app, which serves Traffic Ops endpoints as they're written (currently, just monitoring.json), and reverse-proxies everything else to the old Perl Traffic Ops.

Includes RPM and Service files, to deploy it alongside the old TO.

This can be trivially deployed alongside the old TO with no config (Puppet) changes. It reads the old TO config, and Postinstall sets the new port it needs.

---
Things left before this can be merged:
- [x] Configurable Logging (/traffic_monitor_golang/common/log)
- [x] Unit Tests
- [x] Move to `/traffic_ops` dir for build.sh
- [x] Golang Perl Config parser
- [x] Move `tocookie` out of experimental
- [x] Add header (via wrapper) to Golang endpoints
- [x] Add info to tocookie, indicating Golang created it
- [x] Add Capability to RPM to run unprivileged service on privileged port
- [x] CORS headers
- [x] Routes for `.json` and without
- [x] Log to access.log identical to Perl TO
- [x] Logrotate new log files
- [x] Documentation
- [x] Postinstall adding new high port for old TO
- [x] Configurable max database connections

### CODE REVIEW

**Comment by dewrich** [comment]
I'm not sure why you would couple the Golang Microservices with the Traffic Op Perl Monolith.  If anything you should begin a new "rpm" that only contains the Microservices (which brings up a question about should the MS's be separated into different rpms, or even RPMs at all?)

**Comment by rob05c** [comment]
The reason is to make it as operationally simple as possible. That's the only reason. We can make more RPMs, microservices, etc after we've moved away from Perl. I have no objection to microservices, I'm just afraid if we try to do both at once, we'll never get it done.

**Comment by dewrich** [comment]
I agree with simple, but I also don't want to "couple" the two.  I don't find it egregious, to implement them as a "sidecar" rpm.  We also might want to scale them differently because they are more efficient than Perl.  If at one point we develop enough functionality we want to retire TO Perl, then we will have to unravel the coupling at that point in time.  

If the two are separate then it'll be easy to just "uninstall" the TO Monolith.

**Comment by rob05c** [comment]
It should be easy to retire. Once all endpoints are in Golang, simply delete the Perl from the RPM and Service files.

**Comment by dewrich** [comment]
Also, should `traffic_ops/traffic_ops_golang/auth.go` be in `traffic_ops/experimental/traffic_ops_golang`?  We haven't gotten any buy in from the community to put that code into `traffic_ops` proper yet.

**Comment by rob05c** [comment]
Right, it _will_ cause build changes, don't merge this until we do get consensus from the community on the mailing list (I sent an email to `users` and forwarded to `dev`, "Traffic Ops Golang Migration Proposal".

**Comment by alficles** [comment]
From a user's perspective, this creates some extra (as-yet-poorly-documented) obligations for configuration. An RPM upgrade should take care of itself. There should be enough information for the upgrade to perform all the config changes necessary to maintain nearly exactly the same set of functionality users have today.

You'll need to allocate a new, high, port, but that's a fairly reasonable requirement, easily met. Notify the user via a message during upgrade and I think it would be ok.

I'd love to see this merged; I think it's the right direction. I think we should attend a bit more carefully to the user experience, though. We don't want people afraid to upgrade TO because we break their installs.

**Comment by rob05c** [comment]
> upgrade to perform all the config changes necessary to maintain
> allocate a new, high, port
> Notify the user via a message during upgrade

@alficles Agree, will do. It'll take me a bit though, I'm not an RPM Wizard.

**Comment by dneuman64** [comment]
It would be nice to see some documentation included with this PR before it is merged. 
Also, I don't see any tests, it would be nice if there were some tests included as well.

**Comment by dewrich** [comment]
So, instead of using your own config, why not just use the existing `/opt/traffic_ops/app/conf/production/database.conf`.  Then you wouldn't have to prefix each database property with `db_`.  Also, postinstall sets up that file after the installer Q&A, so then you wouldn't have to touch poinstall

{
  "port": "",
  "db_user": "bill",
  "db_pass": "thelizard",
  "db_server": "db.trafficops.example.net",
  "db_name": "trafficops",
  "db_ssl": true,
  "to_secret": "walrus",
  "to_url": "https://trafficops.example.net:60443",
  "no_auth": false,
  "cert_path": "/opt/traffic_ops/cert",
  "key_path": "/opt/traffic_ops/key"
}
[root@localhost init.d]# cat /opt/traffic_ops/app/conf/production/database.conf
{
   "password" : "twelve",
   "user" : "traffic_ops",
   "type" : "Pg",
   "hostname" : "localhost",
   "description" : "Pg database on localhost:5432",
   "port" : "5432",
   "dbname" : "traffic_ops"
}

If you need properties that aren't in that config then you could use and groom your own

**Comment by dewrich** [comment]
I rolled an RPM to see how it could work, and I'm not seeing the process running.  Which brought up my next question, where do the logs for troubleshooting go?

**Comment by rob05c** [comment]
If you didn't set up the config, it will fail to start.

Right now, failure to start will be logged to the SystemD service log. I'll make log locations part of the config, consistent with our existing apps.

**Comment by rob05c** [comment]
Putting a reminder here: Also need to make setting up the config part of Postinstall.

**Comment by rob05c** [comment]
> so then you wouldn't have to touch poinstall
If we want to automatically set up the config, we're going to have to touch postinstall, if nothing else, to determine the new port to serve old-TO on.

**Comment by rob05c** [comment]
@dneuman64 I agree, I'll add docs and tests.

**Comment by rob05c** [comment]
We're also going to need a Perl serialized hash parser in Go, for `cdn.conf`. It _must_ be modular, so replacing it with JSON is easy once Perl TO is dead. Tight coupling with Perl hash configs is not acceptable.

**Comment by rob05c** [comment]
Things to do before this can be merged:
- [ ] Configurable Logging (/traffic_monitor_golang/common/log)
- [ ] Tests
- [ ] Documentation
- [ ] Postinstall adding new high port for old TO
- [x] Move to root dir for build_all.sh
- [ ] Golang Perl Config parser

**Comment by rob05c** [comment]
Need to add `setcap 'cap_net_bind_service=+ep' /opt/traffic_ops/traffic_ops_golang` to the RPM, and a dependency on libcap2, so the Golang service can run as a user and serve 443. See https://stackoverflow.com/a/414258/292623

**Comment by elsloo** [comment]
@rob05c awesome, looking forward to seeing how that works, as we might want to adopt that approach for any/all of our golang components that need to open ports like this.

File: traffic_ops/traffic_ops_golang/config.go [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+	if cfg, err = ParseConfig(cfg); err != nil {
+		return Config{}, err
+	}
+
+	return cfg, nil
+}
+
+// ParseConfig validates required fields, and parses non-JSON types
+func ParseConfig(cfg Config) (Config, error) {
```
Comment (dewrich) [inline traffic_ops/traffic_ops_golang/config.go]: Feedback I got from the Traffic Logs Generator was each config error should be listed at once.  It's painful for the developer to fix the config problems one by one when they could just look at an error list

File: traffic_ops/traffic_ops_golang/perlconfig.go [hunk 2]
```diff
@@ -0,0 +1,288 @@
...
+	return cfg, nil
+}
+
+func getPort(obj map[string]interface{}) (string, error) {
+	portStrI, ok := obj["traffic_ops_golang_port"]
+	if !ok {
+		return "", fmt.Errorf("missing traffic_ops_golang_port key")
+	}
+	portStr, ok := portStrI.(string)
+	if !ok {
```
Comment (dewrich) [inline traffic_ops/traffic_ops_golang/perlconfig.go]: Shouldn't the port be defined as an "integer"?

Comment (rob05c) [inline traffic_ops/traffic_ops_golang/perlconfig.go]: It's a string just because it was easier to work with in the code, e.g. the Go HTTP server takes a string. I can make it an `int` or `uint` if you want. I'd rather not `uint16` though, it's unusual, and even if performance mattered it's not any faster on a 64-bit processor.

File: traffic_ops/traffic_ops_golang/config.go (hunk 1 above)
Comment (rob05c) [inline traffic_ops/traffic_ops_golang/config.go]: Done.

**Comment by dewrich** [comment]
Can we call `perl_access.log` `traffic_ops_access.log`?

**Comment by dewrich** [comment]
Looks like the latest changes broke the Mojolicious Cookie 

image

**Comment by alficles** [comment]
@dewrich `traffic_ops_access.log` is a bit ambiguous, since the whole construct, go+perl, is Traffic Ops. It should be something that clearly indicates it a log for the perl component.

**Comment by dewrich** [comment]
If we plan on making this a "seamless" deployment how does the `to_url` get configured from traffic_ops_golang.config?  That property doesn't exist in the `cdn.conf`

**Comment by rob05c** [comment]
It reads the port from `cdn.conf` and sets it to `127.0.0.1:port`

File: docs/source/admin/traffic_ops/configuration.rst [hunk 3]
```diff
@@ -296,5 +296,8 @@ This is a standard kickstart formatted file that the generate ISO process uses t
...
 
-
+By default, the postinstall script configures the Go application to behave and transparently serve as the old Perl Traffic Ops did in previous versions. This includes reading the old ``cdn.conf`` and ``database.conf`` config files, and logging to the old ``access.log`` location. However, if you wish to customize the Go Traffic Ops application, you can do so by running it with the ``-oldcfg=false`` argument. By default, it will then look for a config file in ``/opt/traffic_ops/conf/traffic_ops_golang.json``. The new config file location may also be customized via the ``-cfg`` flag. A sample config file is installed by the RPM at ``/opt/traffic_ops/conf/traffic_ops_golang.json``. If you wish to run the new Go Traffic Ops application as a service with a new config file, the ``-oldcfg=false`` and  ``-cfg`` flags may be added to the ``start`` function in the service file, located by default at ``etc/init.d/traffic_ops``.
```
Comment (dewrich) [inline docs/source/admin/traffic_ops/configuration.rst]: Should this `/opt/traffic_ops/conf/traffic_ops_golang.json` be `/opt/traffic_ops/conf/traffic_ops_golang.config`?

File: traffic_monitor_golang/common/log/log.go [hunk 4]
```diff
@@ -95,6 +97,11 @@ func Eventf(t time.Time, format string, v ...interface{}) {
 	Event.Printf("%.3f %s", float64(t.Unix())+(float64(t.Nanosecond())/1e9), fmt.Sprintf(format, v...))
 }
 
+// EventfRaw writes to the event log with no prefix.
+func EventfRaw(format string, v ...interface{}) {
```
Comment (dewrich) [inline traffic_monitor_golang/common/log/log.go]: Should this function be called `AccessfRaw`instead of `EventfRaw`?  Or any of the other functions that write to the `access.log`?

File: traffic_monitor_golang/traffic_monitor/config/config.go [hunk 5]
```diff
@@ -21,7 +21,6 @@ package config
```
Comment (dewrich) [inline traffic_monitor_golang/traffic_monitor/config/config.go]: Why are Traffic Monitor changes in this PR?

File: traffic_ops/traffic_ops_golang/monitoring.go [hunk 6]
```diff
@@ -1,90 +1,45 @@
-// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (dewrich) [inline traffic_ops/traffic_ops_golang/monitoring.go]: Same comment as above, why are monitoring changes in this PR?

File: traffic_ops/traffic_ops_golang/perlconfig.go (hunk 2 above)
Comment (dewrich) [inline traffic_ops/traffic_ops_golang/perlconfig.go]: Was wondering why the decision was made to use a string for a port.  Not what I usually see in config files

File: traffic_ops/traffic_ops_golang/routes.go [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+
+// GetRoutes returns the map of regex routes, and a catchall route for when no regex matches.
+func GetRoutes(d ServerData) (map[string]RegexHandlerFunc, http.Handler, error) {
+	privLevelStmt, err := preparePrivLevelStmt(d.DB)
+	if err != nil {
+		return nil, nil, fmt.Errorf("Error preparing db priv level query: ", err)
+	}
+
+	return map[string]RegexHandlerFunc{
+		"api/1.2/cdns/{cdn}/configs/monitoring":      wrapHeaders(wrapAuth(monitoringHandler(d.DB), d.Insecure, d.TOSecret, privLevelStmt, MonitoringPrivLevel)),
```
Comment (dewrich) [inline traffic_ops/traffic_ops_golang/routes.go]: Can we figure out how to make the routes easier to read?  When the routes start to grow to what they are today, it'll be difficult to read through them in this format.

File: docs/source/admin/traffic_ops/configuration.rst (hunk 3 above)
Comment (rob05c) [inline docs/source/admin/traffic_ops/configuration.rst]: `conf/config` seemed redundant, whereas `.json` immediately tells anyone looking at it what the format is.

File: traffic_monitor_golang/common/log/log.go (hunk 4 above)
Comment (rob05c) [inline traffic_monitor_golang/common/log/log.go]: No, it's using the Event logger for `access.log`. Which seemed to make sense, accesses are events for this app.

File: traffic_monitor_golang/traffic_monitor/config/config.go (hunk 5 above)
Comment (rob05c) [inline traffic_monitor_golang/traffic_monitor/config/config.go]: Because the `log` package is used by both, and the `log.InitCfg` was changed to take an interface, requiring the Monitor `Config` be given that interface's functions.

File: traffic_ops/traffic_ops_golang/monitoring.go (hunk 6 above)
Comment (rob05c) [inline traffic_ops/traffic_ops_golang/monitoring.go]: This isn't a monitoring change, this is the `/monitoring` endpoint in Traffic Ops.

File: traffic_ops/traffic_ops_golang/perlconfig.go (hunk 2 above)
Comment (alficles) [inline traffic_ops/traffic_ops_golang/perlconfig.go]: The reason Go uses strings for ports is to support things like "http" instead of "80". In general, you shouldn't use `strconv.Atoi` to convert port strings. Take a look at https://golang.org/pkg/net/#LookupPort

File: traffic_ops/traffic_ops_golang/perlhash.go [hunk 8]
```diff
@@ -0,0 +1,249 @@
...
+			}
+			obj[key] = v
+		case s[0] == '\'':
+			v := ""
+			v, s, err = getStr(s)
+			if err != nil {
+				return obj, "", fmt.Errorf("Error getting string value after key %v: %v", key, err)
+			}
+			obj[key] = v
+		case unicode.IsDigit(rune(s[0])):
```
Comment (alficles) [inline traffic_ops/traffic_ops_golang/perlhash.go]: You probably know this, but this whole thing will break if the user is using high unicode digit characters as indexes in their array. (Like `৩`, a Bengali three.) You probably don't care, but I feel I should mention it, in case you're aware of a corner case where it matters.

File: traffic_ops/traffic_ops_golang/routes.go (hunk 7 above)
Comment (alficles) [inline traffic_ops/traffic_ops_golang/routes.go]: It's a good point. If you start getting a lot of them, it might get a bit tricky. I'm not sure I see a simple way around it. It already uses wrapper functions to efficiently put the required data all together. Do you have a better suggestion for a way to write it more clearly?

File: traffic_ops/traffic_ops_golang/vendor/gopkg.in/DATA-DOG/go-sqlmock/LICENSE [hunk 9]
```diff
@@ -0,0 +1,28 @@
+The three clause BSD license (http://en.wikipedia.org/wiki/BSD_licenses)
+
```
Comment (alficles) [inline traffic_ops/traffic_ops_golang/vendor/gopkg.in/DATA-DOG/go-sqlmock/LICENSE]: This license is fine, but I don't think I saw an appropriate corresponding entry in the top-level LICENSE file that references this and makes people aware that we've got a non-Apache license here. That definitely has to happen, unless I missed it.

File: traffic_ops/traffic_ops_golang/perlhash.go (hunk 8 above)
Comment (rob05c) [inline traffic_ops/traffic_ops_golang/perlhash.go]: Eh, I care, but it'd be very expensive time-wise to fix properly. I'm sure there are other cases (in fact, we know it doesn't support quoted keys, or double-quoted values). But it works for the Perl config as generated by Postinstall. And since it's only temporary until Perl goes away, I'm voting we don't further delay Step 0 of the migration.

If we hit issues in the future, we can always go back and make the parser more robust.

File: docs/source/admin/traffic_ops/configuration.rst (hunk 3 above)
Comment (dewrich) [inline docs/source/admin/traffic_ops/configuration.rst]: where do I find this `/opt/traffic_ops/conf/traffic_ops_golang.json` I only see `/opt/traffic_ops/conf/traffic_ops_golang.config` in the project, which is why I commented on it being documented as `.json`

**Comment by dewrich** [comment]
As I was performance testing I noticed we don't have any "gates" on the number of concurrent connections coming into the monitoring service.  As mentioned, the following database pooling parameters need to be implemented and configurable:

`db.SetMaxOpenConns(n int)`
`db.SetMaxIdleConns(n int)`

and maybe this one?

`db.SetConnMaxLifetime(d time.Duration)`

File: docs/source/admin/traffic_ops/configuration.rst (hunk 3 above)
Comment (rob05c) [inline docs/source/admin/traffic_ops/configuration.rst]: Ah, you're right. I meant to rename the file and forgot. I'll change the documentation.

**Comment by rob05c** [comment]
I'll make the config for `MaxOpenConns`. I looked in the Go source, and Go will block if it hits that limit, rather than asynchronously returning an error. Which should be fine.

I'd vote we omit `MaxIdleConns` and `ConnMaxLifetime`, until we find a need. `MaxIdleConns` is strictly less than `MaxOpenConns`, and it shouldn't hurt to leave them around, especially for a web service that continuously uses them. It's not reasonable for a SQL server to have a max connection lifetime, web services keeping connections forever is typical, so leaving that infinite should be fine too.

Every additional config is more operations cost to deploy, so we should keep configs to a minimum until someone has a need.

**Comment by dewrich** [comment]
My main concern with fine grained control over database connections is that this service will co-exist with Mojolicious TO database connections, so we have to make sure they both can co-exist in the short term without adverse effects.  

Since Golang is more efficient with connection use, I want the ability to carve out a smaller pool of connections to minimize Postgres "out of connections" errors.  As long as controlling "Idle" connections doesn't affect this goal, I'm ok with excluding `db.SetMaxIdleConns(n int)`.

File: traffic_monitor_golang/traffic_monitor/config/config.go (hunk 5 above)
Comment (dewrich) [inline traffic_monitor_golang/traffic_monitor/config/config.go]: I disagree with TM changes being in the PR because one of the dependencies changed, the smaller we can keep this PR the easier it is to merge (and less conflicts) we have to deal with

File: traffic_ops/traffic_ops_golang/monitoring.go (hunk 6 above)
Comment (dewrich) [inline traffic_ops/traffic_ops_golang/monitoring.go]: sorry I got lost when I was popping back and forth for this review, I see what you mean now

File: traffic_monitor_golang/traffic_monitor/config/config.go (hunk 5 above)
Comment (rob05c) [inline traffic_monitor_golang/traffic_monitor/config/config.go]: The config changes have been moved into https://github.com/apache/incubator-trafficcontrol/pull/620 which has been merged. I'll rebase this to remove them and resolve the conflicts.

Comment (alficles) [inline traffic_monitor_golang/traffic_monitor/config/config.go]: Should the TM changes go in a separate PR on which this one depends, or are you suggesting we branch the log package and keep a separate one for TM and TO?

Comment (alficles) [inline traffic_monitor_golang/traffic_monitor/config/config.go]: Heh, I'm too slow. Apparently a second PR is the right answer. :)

File: traffic_ops/traffic_ops_golang/vendor/gopkg.in/DATA-DOG/go-sqlmock/LICENSE (hunk 9 above)
Comment (rob05c) [inline traffic_ops/traffic_ops_golang/vendor/gopkg.in/DATA-DOG/go-sqlmock/LICENSE]: Done.

File: docs/source/admin/traffic_ops/configuration.rst (hunk 3 above)
Comment (rob05c) [inline docs/source/admin/traffic_ops/configuration.rst]: Fixed.

File: traffic_ops/install/bin/_postinstall [hunk 10]
```diff
@@ -183,6 +183,22 @@ sub generateDbConf {
     return \%todbconf;
```
Comment (alficles) [inline traffic_ops/install/bin/_postinstall]: Did you intend to strip execute permissions from this?

Comment (rob05c) [inline traffic_ops/install/bin/_postinstall]: Fixed.

File: traffic_ops/install/bin/_postinstall [hunk 11]
```diff
@@ -183,6 +183,22 @@ sub generateDbConf {
...
 }
 
+sub setCdnConfGoPort {
+    my $cdnConf = shift;
+    if ( exists $cdnConf->{traffic_ops_golang_port} ) {
```
Comment (dewrich) [inline traffic_ops/install/bin/_postinstall]: This "if statement" needs to be removed because if you run postinstall the second time, then `traffic_ops_golang_port` gets removed because there is an assumption that the cdn.conf will be always regenerated by postinstall

File: traffic_ops/install/bin/_postinstall [hunk 12]
```diff
@@ -218,7 +234,9 @@ sub generateCdnConf {
             $#secrets = $cdnConfiguration{keepSecrets} - 1;
         }
     }
+    $cdnConf = setCdnConfGoPort($cdnConf);
```
Comment (dewrich) [inline traffic_ops/install/bin/_postinstall]: Need another subroutine to account for setting the `max_db_connections` in postinstall (or figuring out reasonable defaults) as discussed

File: traffic_ops/install/bin/_postinstall (hunk 11 above)
Comment (alficles) [inline traffic_ops/install/bin/_postinstall]: I'm confused as to when the port value might be removed. This should always add the port, if it's missing. If the config is regenerated, it might lose the value, but I'm not seeing a code path here that causes the port value to go missing.

File: traffic_ops/install/bin/_postinstall (hunk 12 above)
Comment (alficles) [inline traffic_ops/install/bin/_postinstall]: What's the default max_db_connections for the Perl version?

**Comment by asfgit** [comment]
Can one of the admins verify this patch?

**Comment by rob05c** [comment]
Must be merged after #786 which fixes an ORT bug this exposes.

**Comment by dewrich** [comment]
Just ran a JSON comparison tool against the Golang v. Perl response format and found that they are different.  Unfortunately, this can't change until we rev the API.

Golang
+       hashId: "xxxx"
+       hostName: "xxxx"
+       interfaceName: "xxxx"

Perl
-        hashid: "xxxx"
-        hostname: "xxxx"
-        interfacename: "xxxx"

**Comment by dewrich** [comment]
Per our discussion: 

Mojo HTTP Header is sending this back when the Golang proxy does not.

`Cache-Control: no-cache, no-store, max-age=0, must-revalidate`

**Comment by dewrich** [comment]
This Mojo HTTP Header is also coming back and I don't see it in the Golang HTTP header

`Connection: keep-alive`

**Comment by dewrich** [comment]
Test cases failed:
```
--- FAIL: TestGetCDNConf (0.00s)
	perlconfig_test.go:84: expected {HTTPPort:443 DBUser: DBPass: DBServer: DBDB: DBSSL:false TOSecret:walrus TOURLStr:https://127.0.0.1:60443 TOURL:https://127.0.0.1:60443 Insecure:false CertPath:/etc/pki/tls/certs/localhost.crt KeyPath:/etc/pki/tls/private/localhost.key MaxDBConnections:0 LogLocationError: LogLocationWarning: LogLocationInfo: LogLocationDebug: LogLocationEvent:} actual {HTTPPort:443 DBUser: DBPass: DBServer: DBDB: DBSSL:false TOSecret:walrus TOURLStr:https://127.0.0.1:60443 TOURL:https://127.0.0.1:60443 Insecure:false CertPath:/etc/pki/tls/certs/localhost.crt KeyPath:/etc/pki/tls/private/localhost.key MaxDBConnections:50 LogLocationError: LogLocationWarning: LogLocationInfo: LogLocationDebug: LogLocationEvent:}```

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/22/

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/23/

**Comment by dewrich** [comment]
"test this please"

**Comment by asfgit** [comment]
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/incubator-trafficcontrol-PR/32/
